File: src/algebra1_quadratic_formula_easy.py
"""
algebra1 - quadratic_formula (easy)
Generated: 2026-02-22T04:01:47.330615
"""

from problem_utils import *

def generate():
    difficulty_choice = randint(1, 3)

    if difficulty_choice == 1:
        # Easiest: a=1, integer roots, wide range of root pairs
        r1 = randint(-8, 8)
        r2 = randint(-8, 8)
        # Allow equal roots for variety
        a_val = 1
        b_val = -(r1 + r2)
        c_val = r1 * r2

        equation = Eq(x**2 + b_val*x + c_val, 0)
        disc = b_val**2 - 4*a_val*c_val
        sol1 = (-b_val + sqrt(disc)) / 2
        sol2 = (-b_val - sqrt(disc)) / 2
        ans = FiniteSet(sol1, sol2)

        return problem(
            question=f"Solve using the quadratic formula: ${latex(equation)}$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="algebra1/quadratic_formula",
            solution=steps(
                f"$a={a_val}$, $b={b_val}$, $c={c_val}$",
                f"$x = \\frac{{-({b_val}) \\pm \\sqrt{{({b_val})^2 - 4(1)({c_val})}}}}{{2}}$",
                f"$x = \\frac{{{-b_val} \\pm \\sqrt{{{disc}}}}}{{2}}$",
                f"$x = \\frac{{{-b_val} \\pm {latex(sqrt(disc))}}}{{2}}$",
                f"$x = {latex(sol1)}$ or $x = {latex(sol2)}$"
            ),
            calculator="scientific"
        )

    elif difficulty_choice == 2:
        # Medium-easy: a=1, wider range of roots including negative
        r1 = randint(-10, 10)
        r2 = randint(-10, 10)
        a_val = 1
        b_val = -(r1 + r2)
        c_val = r1 * r2

        equation = Eq(x**2 + b_val*x + c_val, 0)
        disc = b_val**2 - 4*a_val*c_val
        sol1 = (-b_val + sqrt(disc)) / 2
        sol2 = (-b_val - sqrt(disc)) / 2
        ans = FiniteSet(sol1, sol2)

        return problem(
            question=f"Use the quadratic formula to solve: ${latex(equation)}$",
            answer=ans,
            difficulty=(1100, 1200),
            topic="algebra1/quadratic_formula",
            solution=steps(
                f"$a={a_val}$, $b={b_val}$, $c={c_val}$",
                f"Discriminant: $b^2-4ac = ({b_val})^2 - 4({a_val})({c_val}) = {disc}$",
                f"$x = \\frac{{{-b_val} \\pm \\sqrt{{{disc}}}}}{{2}}$",
                f"$x = {latex(sol1)}$ or $x = {latex(sol2)}$"
            ),
            calculator="scientific"
        )

    else:
        # Harder in easy range: a∈{1..5}, b∈{-10..10}, c∈{-20..20} with perfect square disc
        a_val = choice([2, 3, 4, 5])
        r1 = randint(-5, 5)
        r2 = randint(-5, 5)
        b_val = -a_val * (r1 + r2)
        c_val = a_val * r1 * r2

        equation = Eq(a_val*x**2 + b_val*x + c_val, 0)
        disc = b_val**2 - 4*a_val*c_val
        sol1 = (-b_val + sqrt(disc)) / (2*a_val)
        sol2 = (-b_val - sqrt(disc)) / (2*a_val)
        ans = FiniteSet(sol1, sol2)

        return problem(
            question=f"Solve using the quadratic formula: ${latex(equation)}$",
            answer=ans,
            difficulty=(1200, 1300),
            topic="algebra1/quadratic_formula",
            solution=steps(
                f"$a={a_val}$, $b={b_val}$, $c={c_val}$",
                f"Discriminant: $({b_val})^2 - 4({a_val})({c_val}) = {disc}$",
                f"$x = \\frac{{-({b_val}) \\pm \\sqrt{{{disc}}}}}{{2({a_val})}}$",
                f"$x = \\frac{{{-b_val} \\pm {latex(sqrt(disc))}}}{{{2*a_val}}}$",
                f"$x = {latex(sol1)}$ or $x = {latex(sol2)}$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/algebra1_quadratic_formula_easy.jl
# algebra1 - quadratic_formula (easy)
# Generated: 2026-03-08T20:00:49.495174

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra1/quadratic_formula")
    
    # Pick random roots and construct quadratic backward
    r1 = randint(-12, 12)
    r2 = randint(-12, 12)
    a = nonzero(-3, 3)
    
    # Expand to get standard form
    expr = expand(a * (x - r1) * (x - r2))
    
    # Extract coefficients for solution display
    # expr = a*x^2 - a*(r1+r2)*x + a*r1*r2
    coeff_a = a
    coeff_b = -a * (r1 + r2)
    coeff_c = a * r1 * r2
    
    # Discriminant for solution verification
    discriminant = coeff_b^2 - 4*coeff_a*coeff_c
    
    # Sort roots for consistent answer format
    roots = sort([r1, r2])
    
    problem(
        question = "Solve using the quadratic formula: \$$(tex(expr)) = 0\$",
        answer = fmt_set(Set([r1, r2])),
        difficulty = (700, 1200),
        answer_type = "set",
        solution = steps(
            sol("Given", expr ~ 0),
            "Standard form: \$$(coeff_a)x^2 + $(coeff_b)x + $(coeff_c) = 0\$",
            "Apply quadratic formula: \$x = \\frac{-b \\pm \\sqrt{b^2-4ac}}{2a}\$",
            "Here \$a=$(coeff_a)\$, \$b=$(coeff_b)\$, \$c=$(coeff_c)\$",
            "Discriminant: \$\\Delta = $(coeff_b)^2 - 4($(coeff_a))($(coeff_c)) = $(discriminant)\$",
            sol("x = \\frac{-$(coeff_b) \\pm \\sqrt{$(discriminant)}}{$(2*coeff_a)}"),
            "Solutions: \$x = $(roots[1])\$ and \$x = $(roots[2])\$",
            sol("Answer", fmt_set(Set([r1, r2])))
        ),
        time = 120,
        calculator = "scientific"
    )
end